core/src/main/java/org/apache/iceberg/IncrementalFileCleanup.java

```diff
@@ -327,4 +342,34 @@ private Set<String> findFilesToDelete(
 
     return filesToDelete;
   }
+
+  private boolean isSafeToDelete(
+      ManifestEntry<?> entry,
+      Map<Long, Long> validSnapshotIdToSequenceNumberMap,
+      Map<Long, Long> expiredSnapshotIdToSequenceNumberMap) {
+    if (validSnapshotIdToSequenceNumberMap.containsKey(entry.snapshotId())
+        || !expiredSnapshotIdToSequenceNumberMap.containsKey(entry.snapshotId())) {
+      return false;
+    }
+
+    // The file in DELETE entry can be deleted if there are no Snapshots older than
+    // this one
+    if (validSnapshotIdToSequenceNumberMap.keySet().stream()
+        .noneMatch(snapshotId -> snapshotId < entry.snapshotId())) {
+      return true;
```

Comment:
> Snapshot IDs are guaranteed to be unique but not necessarily monotonically increasing (in fact by default in the reference implementation it's random.

Thanks for your reminder, sorry for misleading by `TestTableOperations.newSnapshotId()`. Comparing the `snapshotId` is definitely wrong and I will fix it.

> I think you'll just need to check that there's no parent snapshot for each of the snapshots in the valid snapshots map.

Checking the parent snapshot may also encounter some problems, for example, we may have already expired some specified snapshots including the parent of this snapshot before. In this case, the inheritance chain between snapshots has been broken. So it seems that we shouldn't rely on it as it may lead in wrong result.

For example, if we have a table with actions as follows:

```
table.newAppend().appendFile(FILE_A).commit;	// snapshotID1
table.newAppend().appendFile(FILE_B).commit;	// snapshotID2
table.newDelete().deleteFile(FILE_A).commit;	// snapshotID3
table.newAppend().appendFile(FILE_A2).commit;	// snapshotID4
```

And we have already expired specified snapshot `snapshotID2` before.

Now when we trying to expire snapshot `snapshotID3` using the policy of `IncrementalFileCleanup`, `FILE_A` would be wrongly cleaned up if we rely on checking the parent snapshot.

Reply:
Ah good point, yes the inheritance chain isn't guaranteed since the parent also could've been expired. 
OK I'll need to think about this a bit more deeply then. On the surface just seems like if someone is expiring a specific snapshot where there's a deleted entry we may want to do the reachability analysis

Reply:
> Or potentially since we have the invariant here that there's only one ref at this point, we can just validate that there are no ancestors at all.

Thanks for providing this information. Based on the invariant you mentioned, we can make some useful decisions rely on the inheritance chain. 

Firstly, if we find that the parent snapshot is still valid, we can determine directly that the file in DELETED entry cannot be safely deleted. It must exist in the parent snapshot.

Secondly, although we cannot determine whether all ancestors no longer exist by judging whether there all still valid ancestors, but conversely, we can confirm a situation that all the ancestors are expired in the same operation through checking the expired snapshots collection. This covers a narrower range, but the situation is quite common (due to the `expire older than` operation).

The advantage of rules relying on inheritance chain is that they can support both v1 and v2 tables.

And if the rules relying on inheritance chain do not work, we can take advantage of file sequence number to make some decisions for only v2 tables like what we do now, or maybe could figure out some other rules to do more intelligent decisions.

I have refactored the code to include the rules mentioned above, please take a look when available. Any informations or ideas will be greatly appreciated.